chromium_src/components/signin/public/identity_manager/identity_manager.cc

```diff
@@ -17,6 +17,11 @@
 namespace signin {
 
 AccountsInCookieJarInfo IdentityManager::GetAccountsInCookieJar() const {
+  if (should_fake_next_get_accounts_in_cookie_jar_for_test_) {
```

Comment:
I'm a bit confused here, if we override what this returns in tests then how do we know these changes actually work? https://github.com/brave/brave-core/pull/25971/files#diff-9cc578b6bc0ff8c20ea911e73c25e9aa392ad37c3a6877bc71b45bbf611ca649

Reply:
yes, it should , and it does.

but when I do override the `BraveSyncServiceImpl::OnPrimaryAccountChanged` and I want to test that my change actually prevents cache id regeneration, I need to simulate the scenario when without my empty override I can see the cache guid regeneration.

But I can't because to make `user_settings.KeepAccountSettingsPrefsOnlyForUsers(hashes);` be invoked, I need to have `accounts_in_cookie_jar_info.accounts_are_fresh == true`
```
void MaybeClearAccountKeyedPreferences(
    signin::IdentityManager* identity_manager,
    const signin::AccountsInCookieJarInfo& accounts_in_cookie_jar_info,
    SyncUserSettingsImpl& user_settings) {
#if !BUILDFLAG(IS_IOS) && !BUILDFLAG(IS_ANDROID)
  if (accounts_in_cookie_jar_info.accounts_are_fresh) {
    // Clear settings for accounts no longer in the cookie jar. On Android
    // and iOS this is done when the account is removed from the OS instead.
    std::vector<signin::GaiaIdHash> hashes =
        base::ToVector(signin::GetAllGaiaIdsForKeyedPreferences(
                           identity_manager, accounts_in_cookie_jar_info),
                       &signin::GaiaIdHash::FromGaiaId);
    user_settings.KeepAccountSettingsPrefsOnlyForUsers(hashes);
  }
#endif  // !BUILDFLAG(IS_IOS) && !BUILDFLAG(IS_ANDROID)
}
```

So @bridiver , according to your questions and confusion, I suppose the correct way would be just to remove all these things :
- `BraveSyncServiceImpl::OnPrimaryAccountChanged` override;
- test for `OnPrimaryAccountChanged`;
- `IdentityManager::FakeGetAccountsInCookieJarForNextCallForTests`
?

Reply:
@bridiver 

>shouldn't this always be empty since we don't use gaia?

and the call stack for the issue when device guid gets re-generated after  logout from GMail:

```
+#2 0x61f515dba9d4 syncer::KeepAccountKeyedPrefValuesOnlyForUsers()
+#3 0x61f515db915e syncer::SyncTransportDataPrefs::KeepAccountSettingsPrefsOnlyForUsers()
+#4 0x61f515e30806 syncer::SyncUserSettingsImpl::KeepAccountSettingsPrefsOnlyForUsers()
+#5 0x61f515ddb97b syncer::(anonymous namespace)::MaybeClearAccountKeyedPreferences()
+#6 0x61f515ddbaa8 syncer::SyncServiceImpl::OnAccountsInCookieUpdatedWithCallback()
+#7 0x61f515ddbc8b syncer::SyncServiceImpl::OnAccountsInCookieUpdated()
+#8 0x61f5139a4d53 signin::IdentityManager::OnGaiaAccountsInCookieUpdated()
+#9 0x61f5139c69f6 base::RepeatingCallback<>::Run()
+#10 0x61f5139c7444 GaiaCookieManagerService::OnListAccountsSuccess()
+#11 0x7dc343c10435 GaiaAuthFetcher::OnListAccountsFetched()
+#12 0x7dc343c0d34b GaiaAuthFetcher::OnURLLoadComplete()
+#13 0x7dc343b894bc base::OnceCallback<>::Run()
+#14 0x7dc343b89ce9 network::(anonymous namespace)::SimpleURLLoaderImpl::FinishWithResult()
+#15 0x7dc343b89e6d network::(anonymous namespace)::SimpleURLLoaderImpl::MaybeComplete()
+#16 0x7dc343b8533f network::(anonymous namespace)::SimpleURLLoaderImpl::OnComplete()
+#17 0x7dc343bd133c network::mojom::URLLoaderClientStubDispatch::Accept()

+#0 0x75e9edd753f2 base::debug::CollectStackTrace()
+#1 0x75e9edd5236a base::debug::StackTrace::StackTrace()
+#2 0x75e9eaf4ed9f GaiaAuthFetcher::CreateAndStartGaiaFetcher()
+#3 0x75e9eaf50735 GaiaAuthFetcher::StartListAccounts()
+#4 0x64f162015095 GaiaCookieManagerService::StartFetchingListAccounts()
+#5 0x64f162016d8b base::internal::Invoker<>::RunOnce()
+#6 0x64f161f74270 base::OnceCallback<>::Run()
+#7 0x64f163df7627 WaitForNetworkCallbackHelperChrome::DelayNetworkCall()
+#8 0x64f163d3f1e2 ChromeSigninClient::DelayNetworkCall()
+#9 0x64f162011cb0 GaiaCookieManagerService::TriggerListAccounts()
+#10 0x64f162012cc0 GaiaCookieManagerService::OnCookieChange()
+#11 0x64f1620b2feb network::mojom::CookieChangeListenerStubDispatch::Accept()

```

Reply:
I think it's fine to leave the overrides, we should just have some comments explaining why we're overriding them

Reply:
@AlexeyBarabash I didn't mean we need to test it like this, I just meant like
```
              signin::PrimaryAccountChangeEvent primary_accountchange_event(
                  signin::PrimaryAccountChangeEvent::State(
                      prev_account_info, signin::ConsentLevel::kSignin),
                  signin::PrimaryAccountChangeEvent::State(),
                  signin_metrics::ProfileSignout::kTest);

              p_this->brave_sync_service_impl()->OnPrimaryAccountChanged(
                  primary_accountchange_event);
```
to make sure it won't set any sync prefs, even though we already know it won't because it is empty.
Not writing it like what upstream code was supposed to work and then try to fulfill the conditions to prove our override works, which is not sustainable and upstream can change any of this in the future.

Reply:
thanks, updated the PR:
https://github.com/brave/brave-core/pull/25971/commits/dc7de3309656d7549a96c1febbd88e5e0a7aa0fb

force-pushed after rebase